File: Python/leetcode_1002.py
from collections import Counter
from typing import List


def common_chars(words: List[str]) -> List[str]:
    common = Counter(words[0])

    for word in words[1:]:
        common &= Counter(word)  # intersection: min counts

    result = []
    for char, freq in common.items():
        result.extend([char] * freq)

    return result


print(common_chars(["cool", "lock", "cook"]))

File: JavaScript/leetcode_1002.js
function commonChars(words) {
    // Helper to count characters in a word
    const countChars = (word) => {
        const count = Array(26).fill(0);
        for (let char of word) {
            count[char.charCodeAt(0) - 97]++;
        }
        return count;
    };

    // Initialize with the count from the first word
    let common = countChars(words[0]);

    // Intersect counts with each next word
    for (let i = 1; i < words.length; i++) {
        const currentCount = countChars(words[i]);
        for (let j = 0; j < 26; j++) {
            common[j] = Math.min(common[j], currentCount[j]);
        }
    }

    // Build result array from final counts
    const result = [];
    for (let i = 0; i < 26; i++) {
        while (common[i] > 0) {
            result.push(String.fromCharCode(i + 97));
            common[i]--;
        }
    }

    return result;
}
